src/transformStateWithClones.js

```diff
@@ -6,8 +6,50 @@
  *
  * @return {Object[]}
  */
+
+const ACTION_TYPES = {
+  add: 'addProperties',
+  remove: 'removeProperties',
+  clear: 'clear',
+};
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+
+  for (const { type, extraData, keysToRemove } of actions) {
+    switch (type) {
+      case ACTION_TYPES.add: {
+        const prevStateCopy = { ...(history.at(-1) ?? state) };
+
+        history.push({ ...prevStateCopy, ...extraData });
```

Comment:
Instead of creating a new state copy inside each case, you could create it once at the beginning of the loop iteration and then modify it accordingly in each case.